File: scripts/create-icons.py
#!/usr/bin/env python3
"""
FoxiFill icon generation script.
Uses PIL (Pillow) to generate PNG icons in required extension sizes.
"""

import os
import sys
import argparse

def create_icons():
    """Generate extension icons."""
    
    # Path configuration
    script_dir = os.path.dirname(os.path.abspath(__file__))
    extension_dir = os.path.dirname(script_dir)
    project_root = os.path.dirname(os.path.dirname(extension_dir))
    
    # Source icon path
    source_icon = os.path.join(project_root, "01-vi", "icon.png")
    icons_dir = os.path.join(extension_dir, "public", "icons")
    required_icon_files = [
        os.path.join(icons_dir, f"icon{size}.png")
        for size in [16, 32, 48, 128]
    ]
    
    print("FoxiFill icon generator")
    print(f"Source icon: {source_icon}")
    print(f"Output directory: {icons_dir}")
    
    # Check source file
    if not os.path.exists(source_icon):
        if all(os.path.exists(icon_file) for icon_file in required_icon_files):
            print("Source icon not found. Existing extension icons are present, skipping regeneration.")
            return True
        print(f"Error: source icon file not found: {source_icon}")
        print("Expected existing icon files in public/icons when building outside the original monorepo.")
        return False

    try:
        from PIL import Image
    except ImportError:
        print("Error: missing Pillow dependency")
        print("Install with: pip install Pillow")
        return False
    
    # Create output directory
    os.makedirs(icons_dir, exist_ok=True)
    
    try:
        # Load source icon
        print("Loading source icon...")
        with Image.open(source_icon) as source_img:
            print(f"   Size: {source_img.size}")
            print(f"   Mode: {source_img.mode}")
            
            # Ensure the image has an alpha channel.
            if source_img.mode != 'RGBA':
                source_img = source_img.convert('RGBA')
            
            # Required extension icon sizes
            sizes = [16, 32, 48, 128]
            
            for size in sizes:
                print(f"Generating {size}x{size} icon...")
                
                # Resize with high-quality resampling.
                resized = source_img.resize((size, size), Image.Resampling.LANCZOS)
                
                # Save PNG
                png_path = os.path.join(icons_dir, f"icon{size}.png")
                resized.save(png_path, "PNG", optimize=True)
                
                # Validate file
                file_size = os.path.getsize(png_path)
                print(f"   Created {png_path} ({file_size} bytes)")
                
                # Also generate a simplified SVG variant.
                svg_content = create_svg_icon(size)
                svg_path = os.path.join(icons_dir, f"icon{size}.svg")
                with open(svg_path, 'w', encoding='utf-8') as f:
                    f.write(svg_content)
                print(f"   Created {svg_path}")
        
        # Copy the original logo as a reference asset.
        try:
            logo_src = os.path.join(project_root, "01-vi", "logo.svg")
            if os.path.exists(logo_src):
                import shutil
                logo_dst = os.path.join(icons_dir, "logo-original.svg")
                shutil.copy2(logo_src, logo_dst)
                print(f"Copied original logo: {logo_dst}")
        except Exception as e:
            print(f"Could not copy logo: {e}")
        
        print("Icon generation completed.")
        print("FoxiFill extension icons are ready.")
        return True
        
    except Exception as e:
        print(f"Error generating icons: {e}")
        return False

def create_svg_icon(size):
    """Create an SVG icon variant."""
    return f'''<svg width="{size}" height="{size}" viewBox="0 0 {size} {size}" xmlns="http://www.w3.org/2000/svg">
    <defs>
        <linearGradient id="grad{size}" x1="0%" y1="0%" x2="100%" y2="100%">
            <stop offset="0%" style="stop-color:#F67B26;stop-opacity:1" />
            <stop offset="50%" style="stop-color:#E56920;stop-opacity:1" />
            <stop offset="100%" style="stop-color:#CC5610;stop-opacity:1" />
        </linearGradient>
        <filter id="shadow{size}" x="-50%" y="-50%" width="200%" height="200%">
            <feDropShadow dx="1" dy="1" stdDeviation="1" flood-color="#00000020"/>
        </filter>
    </defs>
    
    <!-- Rounded background -->
    <rect width="{size}" height="{size}" rx="{size * 0.15}" ry="{size * 0.15}" 
          fill="url(#grad{size})" filter="url(#shadow{size})" />
    
    <!-- FoxiFill mark -->
    <text x="{size * 0.5}" y="{size * 0.65}" 
          font-family="Arial, -apple-system, sans-serif" 
          font-size="{size * 0.4}" 
          font-weight="bold" 
          fill="white" 
          text-anchor="middle" 
          dominant-baseline="middle">F</text>
    
    <!-- Accent dot -->
    <circle cx="{size * 0.8}" cy="{size * 0.2}" r="{size * 0.05}" 
            fill="white" opacity="0.8" />
</svg>'''

def check_dependencies():
    """Check dependencies."""
    try:
        import PIL
        print(f"Pillow version: {PIL.__version__}")
        return True
    except ImportError:
        print("Error: missing Pillow dependency")
        print("Install with: pip install Pillow")
        return False

def main():
    """Main entry point."""
    parser = argparse.ArgumentParser(description='FoxiFill icon generator')
    parser.add_argument('--check', action='store_true', help='Check dependencies')
    args = parser.parse_args()
    
    if args.check:
        return 0 if check_dependencies() else 1

    success = create_icons()
    return 0 if success else 1

if __name__ == "__main__":
    sys.exit(main())

File: scripts/create-icons.js
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Create SVG-based icons for extension packaging.
const createIconSVG = (size) => {
  return `<svg width="${size}" height="${size}" viewBox="0 0 ${size} ${size}" xmlns="http://www.w3.org/2000/svg">
    <defs>
      <linearGradient id="grad1" x1="0%" y1="0%" x2="100%" y2="100%">
        <stop offset="0%" style="stop-color:#F67B26;stop-opacity:1" />
        <stop offset="50%" style="stop-color:#E56920;stop-opacity:1" />
        <stop offset="100%" style="stop-color:#CC5610;stop-opacity:1" />
      </linearGradient>
      <filter id="shadow" x="-50%" y="-50%" width="200%" height="200%">
        <feDropShadow dx="1" dy="1" stdDeviation="1" flood-color="#00000030"/>
      </filter>
    </defs>
    
    <!-- Rounded background -->
    <rect width="${size}" height="${size}" rx="${size * 0.2}" ry="${size * 0.2}" 
          fill="url(#grad1)" filter="url(#shadow)" />
    
    <!-- Fox head outline -->
    <ellipse cx="${size * 0.5}" cy="${size * 0.45}" rx="${size * 0.25}" ry="${size * 0.2}" 
             fill="#FFFFFF" opacity="0.9" />
    
    <!-- Fox ears -->
    <polygon points="${size * 0.35},${size * 0.3} ${size * 0.4},${size * 0.15} ${size * 0.45},${size * 0.3}" 
             fill="#FFFFFF" opacity="0.9" />
    <polygon points="${size * 0.55},${size * 0.3} ${size * 0.6},${size * 0.15} ${size * 0.65},${size * 0.3}" 
             fill="#FFFFFF" opacity="0.9" />
    
    <!-- Letter F -->
    <text x="${size * 0.5}" y="${size * 0.78}" 
          font-family="Arial, sans-serif" 
          font-size="${size * 0.35}" 
          font-weight="bold" 
          fill="#FFFFFF" 
          text-anchor="middle" 
          dominant-baseline="middle">F</text>
    
    <!-- Accent dot -->
    <circle cx="${size * 0.8}" cy="${size * 0.2}" r="${size * 0.06}" 
            fill="#FFFFFF" opacity="0.7" />
  </svg>`;
};

// Create a basic PNG data structure.
const createBasicPNG = (size) => {
  // This simplified fallback keeps the script dependency-free.
  
  // Create RGBA pixel data.
  const orangeColor = [246, 123, 38, 255];
  const whiteColor = [255, 255, 255, 255];
  const transparentColor = [0, 0, 0, 0];
  
  // Create pixel array.
  const pixels = [];
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      // Create rounded rectangle background.
      const isInBounds = 
        (x >= 2 && x < size - 2 && y >= 0 && y < size) ||
        (x >= 0 && x < size && y >= 2 && y < size - 2) ||
        (x >= 1 && x < size - 1 && y >= 1 && y < size - 1);
      
      if (isInBounds) {
        // Add a simplified letter F.
        const isF = 
          (x >= Math.floor(size * 0.3) && x <= Math.floor(size * 0.4) && 
           y >= Math.floor(size * 0.4) && y <= Math.floor(size * 0.8)) ||
          (x >= Math.floor(size * 0.3) && x <= Math.floor(size * 0.6) && 
           y >= Math.floor(size * 0.4) && y <= Math.floor(size * 0.5)) ||
          (x >= Math.floor(size * 0.3) && x <= Math.floor(size * 0.55) && 
           y >= Math.floor(size * 0.6) && y <= Math.floor(size * 0.65));
        
        if (isF) {
          pixels.push(...whiteColor);
        } else {
          pixels.push(...orangeColor);
        }
      } else {
        pixels.push(...transparentColor);
      }
    }
  }
  
  // Return simplified PNG data.
  const pngData = Buffer.from([
    0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A, // PNG signature
    0x00, 0x00, 0x00, 0x0D, // IHDR chunk length
    0x49, 0x48, 0x44, 0x52, // IHDR chunk type
    0x00, 0x00, 0x00, 0x01, // width (1)
    0x00, 0x00, 0x00, 0x01, // height (1)
    0x08, 0x06, // bit depth and color type (RGBA)
    0x00, 0x00, 0x00, // compression, filter, interlace
    0x1F, 0x15, 0xC4, 0x89, // CRC
    0x00, 0x00, 0x00, 0x0B, // IDAT chunk length
    0x49, 0x44, 0x41, 0x54, // IDAT chunk type
    0x78, 0x9C, 0x62, 0xF8, 0x7B, 0x26, 0xFF, 0x00, 0x00, 0x00, 0x05, // compressed orange pixel data
    0x00, 0x01, 0xE2, 0x26, 0x05, 0x9B, // CRC
    0x00, 0x00, 0x00, 0x00, // IEND chunk length
    0x49, 0x45, 0x4E, 0x44, // IEND chunk type
    0xAE, 0x42, 0x60, 0x82  // CRC
  ]);
  
  return pngData;
};

// Create icon directory.
const iconsDir = path.join(__dirname, "../public/icons");
if (!fs.existsSync(iconsDir)) {
  fs.mkdirSync(iconsDir, { recursive: true });
}

console.log("Creating FoxiFill extension icons...");

// Create icons in required sizes.
const sizes = [16, 32, 48, 128];
sizes.forEach((size) => {
  // Create SVG variant.
  const svgContent = createIconSVG(size);
  const svgPath = path.join(iconsDir, `icon${size}.svg`);
  fs.writeFileSync(svgPath, svgContent);
  console.log(`Created SVG: ${svgPath}`);
  
  // Create PNG variant.
  const pngData = createBasicPNG(size);
  const pngPath = path.join(iconsDir, `icon${size}.png`);
  fs.writeFileSync(pngPath, pngData);
  console.log(`Created PNG: ${pngPath}`);
});

// Try to copy the project logo as a reference asset.
try {
  const logoSrcPath = path.join(__dirname, "../../../01-vi/logo.svg");
  const logoDestPath = path.join(iconsDir, "logo-original.svg");
  
  if (fs.existsSync(logoSrcPath)) {
    fs.copyFileSync(logoSrcPath, logoDestPath);
    console.log("Copied original logo for reference");
  }
} catch {
  console.log("Could not copy original logo (not critical)");
}

console.log("All icon files created successfully.");
console.log("FoxiFill is ready to go.");
